Initial Plugin version

rnr opened the pull request:
Added Initial code, github workflows and Tests

rnr committed:
chore: initial code commit (#1)

Co-authored-by: Anton Yarmolenko <37253+rnr@users.noreply.github.com>

rnr committed:
chore: added github workflows (#2)

* chore: added github workflows

* fix: scheme name

* chore: test linter error

* fix: linter error

---------

Co-authored-by: Anton Yarmolenko <37253+rnr@users.noreply.github.com>

rnr committed:
chore: updated readme

rnr committed:
chore: fix readme

rnr committed:
chore: added license, edited readme

rnr committed:
chore: update readme

rnr committed:
chore: updated readme

rnr committed:
chore: Added unit Tests (#3)

* chore: added tests

* chore: fixed warnings

* chore: fixed swiftlint warnings in tests

* chore: added unit_test workflow

---------

Co-authored-by: Anton Yarmolenko <37253+rnr@users.noreply.github.com>

mta452 commented on Sources/EDXMobileAnalytics/Segment/SegmentAnalyticsService.swift:
```diff
@@ -0,0 +1,64 @@
+    }
+    
+    public func identify(id: String, username: String?, email: String?) {
+        guard let email = email, let username = username else { return }
```
If we require `email` and `username`, then I would suggest to add `assertionFailure` in `else` statement to avoid surprises.

mta452 commented on Sources/EDXMobileAnalytics/Segment/SegmentAnalyticsServiceMock.swift:
I would suggest to use an existing solution to generate boilerplate code, such as https://github.com/Kolos65/Mockable

rnr replied:
I don't think we need to add another dependency to generate only one mock file. Perhaps we will add it later.
It also looks like tool you mentioned can't work with inherited protocols. Thank you

mta452 replied:
Alright, I think we can postpone this discussion for now. In that case, it makes sense to move this file to test target only.

rnr replied:
moved. thank you

mta452 commented on Tests/EDXMobileAnalyticsTests/BrazeListenerTests.swift:
```diff
@@ -0,0 +1,72 @@
+            let deepLinkMananger = DeepLinkManagerProtocolMock()
+            let brazeListener = BrazeListener(deepLinkManager: deepLinkMananger)
+            let analyticService = Mirror(reflecting: brazeListener)
+                .descendant("segmentAnalyticService") as? SegmentAnalyticsServiceProtocol
```
This can be difficult to catch in case of refactoring. I would suggest an existing solution, such as: https://github.com/fernandolucheti/TestableMacro

mta452 commented on Sources/EDXMobileAnalytics/EDXMobileAnalytics.swift:
Is this file required?

rnr replied:
No, it's just auto-created. Removing. Thank you

mta452 commented on README.md:
```diff
@@ -1 +1,110 @@
+
+## License
+
+This project is licensed under the MIT License. See the [LICENSE](LICENSE) file for details.
```
The `LICENSE` file points to Apache License 2.0 though.

mta452 commented on .github/workflows/xcodebuild.yml:
build and test can be merged in a single workflow to save some time.

rnr commented on Package.swift:
```diff
@@ -0,0 +1,47 @@
+let package = Package(
+    name: "EDXMobileAnalytics",
+    platforms: [
+        .iOS(.v16)
```
Yes openEdx was moved to iOS 16 in October [here](https://github.com/openedx/openedx-app-ios/commit/c632a014c3400cc190b45dda4520912ec42e1983#diff-7b5c135da6260dff4ba909b8427a2b13400744228119797c6493a76d3ec22c26R772) - before the plugin architecture was implemented

mta452 replied:
But we are still supporting iOS 15 in our app and as far as I understand, this package will be included in our app which will force the app to have deployment target of iOS 16 at least. I think it should be discussed in a Slack thread with wider audience so that everyone from business, product and dev side is aligned on the decision.

rnr replied:
Yes, this package will be included in the edX app, but after upstream development (which is iOS 16 and swift 6) will only be merged into the 2u/develop marketplace. And yes, we could discuss this at the next tech meeting. Thanks.

rnr committed:
chore: address feedback

Co-Authored-By: Anton Yarmolenko <37253+rnr@users.noreply.github.com>

rnr commented on Sources/EDXMobileAnalytics/Braze/BrazeProvider.swift:
```diff
@@ -0,0 +1,34 @@
+                additionalConfiguration: { configuration in
+                    configuration.logger.level = .info
+                }, additionalSetup: { braze in
+                    braze.notifications.register(deviceToken: deviceToken)
```
Yes, Braze is a service for push notifications and event tracking, implemented in our case through Segment Analytics Service. We decided not to create a separate repo for these things.

rnr replied:
We had a whole conversation about the name and settled on this name since most of the plugin is analytics (Segment, FullStory will be added, Braze is part analytics). I'm not against any name, but we need to discuss it with the whole team and coordinate with android team and change there as well. I'm not sure it's worth the effort.

mta452 replied:
Here's my take on it. If it only needs to provide analytics services, then the existing name makes full sense. In this case, we should find proper abstraction for Braze. Maybe we don't need to include `BrazeProvider` and `BrazeListener` in this repository at all and handle it internally via a function, say `addBrazePlugin()` in `SegmentAnalyticsService`.

However, if it needs to include other responsibilities as well, such as push notification handling, then we should definitely consider renaming it.

rnr replied:
Since Braze is integrated through Segment, it makes sense to keep it in one repo.
I agree about the plugin naming - it could be more “generic”. For example, `EDXServiceKit` or `EDXEngageKit` you suggested. Lets create a topic in Slack to discuss.

rnr commented on Sources/EDXMobileAnalytics/Braze/BrazeListener.swift:
```diff
@@ -0,0 +1,36 @@
+        guard shouldListenNotification(userinfo: userInfo) else { return }
+        
+        segmentAnalyticService?.receivedRemoteNotification(userInfo: userInfo)
+        deepLinkManager.processLinkFrom(userInfo: userInfo)
```
same as https://github.com/edx/edx-mobile-app-analytics-ios/pull/4#discussion_r1923677766

rnr commented:
@mta452 your feedback was addressed and/or commented.
Ready for another pass. Thank you

mta452 commented:
@rnr I have added minor suggestions on the comments. Other than that it looks good to me. I'll approve it once we reach an understanding.

rnr committed:
chore: moved Mock to test target

rnr commented:
> @rnr I have added minor suggestions on the comments. Other than that it looks good to me. I'll approve it once we reach an understanding.

moved. thank you

bilalawan321 commented on .github/workflows/build_and_test.yml:
```diff
@@ -0,0 +1,41 @@
+    name: Build and Test (iOS 18.1)
+    runs-on: macos-latest
+    env:
+      DEVELOPER_DIR: "/Applications/Xcode_16.1.app/Contents/Developer"
```
The Xcode version (Xcode_16.1.app) is hardcoded. 
Consider using a variable or input parameter to make it easier to update.

rnr replied:
the same here - trying to follow openedx plugins workflows:
https://github.com/openedx/openedx-app-foundation-ios/blob/main/.github/workflows/xcodebuild.yml

bilalawan321 commented:
@rnr I have added some minor suggestions that don’t block the PR from being merged. Great work overall! If these suggestions make sense, they can be addressed in this or in a follow-up PR. 🚀

rnr commented on .github/workflows/swiftlint.yml:
```diff
@@ -0,0 +1,33 @@
+
+      - name: Setup
+        run:
+          xcodes select 16.1
```
I didn't create new one workflow - just use as openedx plugins do
https://github.com/openedx/openedx-app-foundation-ios/blob/main/.github/workflows/swiftlint.yml

rnr committed:
chore: added description

UsamaSadiq committed:
build: update commitlint workflow to use openedx reusable workflow instead (#7)

rnr committed:
feat: Add Fullstory (#6)

* chore: added fullstory service

* chore: update firebase version

* chore: added fullstory tests

* chore: address review feedback

* chore: delete unneeded let

* chore: moved DeepLinkManagerProtocolMock from oexfoundation to here

---------

Co-authored-by: Anton Yarmolenko <37253+rnr@users.noreply.github.com>

rnr committed:
chore: pointed to 1.0.3 foundation

rnr committed:
chore: updated readme file
